thetis/utility.py

```diff
@@ -306,11 +306,64 @@ def comp_tracer_mass_2d(eta, bath, scalar_func):
     :arg bath: bathymetry :class:`Function`
     :arg scalar_func: scalar :class:`Function` to integrate
     """
-
     val = assemble((eta+bath)*scalar_func*dx)
     return val
 
 
+def comp_tracer_total_mass_2d(var, tracer_name):
+    """
+    Computes total tracer mass in the 2D domain accounting for tracer leaving through
+    boundary conditions and source terms
+    :arg var: class:`DiagnosticCallback`; callback object used to input values and
+                                          record output values
+    :arg tracer_name :class:`string` of function name of interest
+    """
+
+    # read in necessary variables from solver object
+    eta = var.solver_obj.fields.elev_2d
+    vel = var.solver_obj.fields.uv_2d
+
+    scalar_func = var.solver_obj.fields[tracer_name]
+
+    # calculate total depth
+    term = var.solver_obj.eq_tracer.terms['HorizontalAdvectionTerm']
+    H = term.get_total_depth(eta)
+
+    # normal
+    n = FacetNormal(var.solver_obj.mesh2d)
+
+    # calculate contribution from tracer leaving boundary
+    boundary_terms = 0
+
+    for bnd_marker in term.boundary_markers:
+        ds_bnd = ds(int(bnd_marker), degree=term.quad_degree)
+        a = -(n[0]*vel[0] + n[1]*vel[1])*H*scalar_func*ds_bnd
+
+        boundary_terms += assemble(a)
```

Comment:
So here you are trying to compute the contribution of source and boundary terms to the total mass. Note that this is not exact: the callback is only called when writing the outputs to disk. So you're essentially assuming the flux is constant over `simulation_export_time`. The fluxes should be evaluated at every time step. And even that would not work if you're using a multi-stage time integrator. In the case of other volume/mass conservation checks I have simply ignored the boundary fluxes and source terms — thus the metrics only make sense in a closed domain. Do you have applications where you actually need to track the mass with open boundaries and source terms?

Reply:
This callback is designed to be called every timestep (see the test_consistency_2d.py for an example) exactly because of the problems you describe. The reason I wanted to track the source term is so that I could track tracer conservation when the bedlevel changes. The bedlevel changes as a function of the source term

Reply:
OK, I see. Is this exact even for Crank-Nicolson? Backward Euler should work.